### ISSUE TITLE
Support newer XPath versions (3.0, 3.1)

### ISSUE THREAD

**Issue by cdanger** [oasis-tcs/xacml-spec#29]
XACML 3.0 only supports XPath 1.0 and 2.0. This is a enhancement proposal to support XPath 3.0 and/or latest 3.1.

The new features in XPath 3.0 / 3.1 are well summarized in [Saxonica's documentation](https://www.saxonica.com/html/documentation12/expressions/index.html). Just to give a few:

- String concatenation operator (||) 
- Simple mapping operator (!)
- _let_ expressions for declaring local variables
- New functions head, tail, for-each, filter, for-each-pair, pi, sin, cos, tan, asin, acos, atan, sqrt, format-integer, and others
- XPath maps and arrays.

**Comment by steven-legg** [comment]
It makes sense to support the latest available version. Support for XPath is optional so including 3.0/3.1 doesn't impose a burden on implementators who ignore XPath.

There is a section in the core spec that discusses the implementation-defined aspects of XPath 2.0. We should also cover XPath 3.0/3.1.

**Comment by cdanger** [comment]
I added a paragraph to my original comment on **JSON processing capabilities introduced in XPath 3.1**, which allows to mimic JSONPath (#5 ) evaluation with XPath syntax!

**Comment by cdanger** [comment]
Should we also deprecate XPath 1.0?

**Comment by humantypo** [comment]
25 years seems like a good run so that makes sense to me. Given that XPath 3.1 has been around for 7 years is it reasonable to ask the same thing about XPath 2.0 in v4?

**Comment by cdanger** [comment]
Well, XPath 2.0 was last updated in 2016 and still widely used as far as I know, so I would not throw it away just yet. Also last time I checked there are not so many XPath 3.x implementations out there. But we have at least one in Java, so no problem there, but other developers in other programming languages might not be so lucky. What's your experience?

**Comment by humantypo** [comment]
My experience is that everything in my world has moved to JSON :)

I did some poking around and it looks like XPath 2 is still very much in use...to the point that it seems pretty clear there isn't a lot of interest in XPath these days.

**Comment by cdanger** [comment]
Indeed, it's time to move that to a separate xacml profile #33 so that the JSON fan base doesn't have to worry too much about that in the core spec :stuck_out_tongue_winking_eye: .

**Comment by cdanger** [comment]
Right, striked that out. So the [XPath profile](#36 ) then. I was too much convinced that the people using JSON for policies would not give a damn about XPath or other XML specifics at all 😉 .

**Comment by steven-legg** [comment]
> Not sure what to do with all these. Should we address them all? How?

XACML 3.0 left most of the (less numerous) XPath 2.0 items implementation defined. I suggest we leave the 3.0/3.1 items implementation defined except where there is a correspondence to a 2.0 item where XACML 3.0 chose a particular behaviour, or where we see that the choice matters to ACAL.

We don't need to have a resolution for every item in the text. Call out the items where we impose a particular behaviour and say that the rest are implementation defined.

**Comment by cdanger** [comment]
Alright, I have left the rest implementation-defined. Fixed in 5819513.
